handlers/utils/compute_utils.go

```diff
@@ -0,0 +1,1186 @@
+package utils
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/engine-api/client"
+	"github.com/docker/engine-api/types"
+	"github.com/docker/engine-api/types/blkiodev"
+	"github.com/docker/engine-api/types/container"
+	"github.com/docker/engine-api/types/filters"
+	"github.com/docker/engine-api/types/strslice"
+	"github.com/docker/go-connections/nat"
+	"github.com/mitchellh/mapstructure"
+	"github.com/rancher/agent/handlers/docker"
+	"github.com/rancher/agent/handlers/marshaller"
+	"github.com/rancher/agent/handlers/progress"
+	"github.com/rancher/agent/model"
+	"github.com/rancher/go-machine-service/events"
+	revents "github.com/rancher/go-machine-service/events"
+	"golang.org/x/net/context"
+	"io/ioutil"
+	urls "net/url"
+	"os"
+	"path"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+func DoInstanceActivate(instance *model.Instance, host *model.Host, progress *progress.Progress) error {
+	if IsNoOp(instance.Data) {
+		return nil
+	}
+	client := docker.GetClient(DefaultVersion)
+
+	imageTag, err := getImageTag(instance)
+	if err != nil {
+		return err
+	}
+	name := instance.UUID
+	instanceName := instance.Name
+	if len(instanceName) > 0 {
+		if ok, _ := regexp.Match("^[a-zA-Z0-9][a-zA-Z0-9_.-]+$", []byte(instanceName)); ok {
+			id := fmt.Sprintf("r-%s", instanceName)
+			_, err1 := client.ContainerInspect(context.Background(), id)
+			if err1 != nil {
+				name = id
+			}
+		}
+	}
+	var createConfig = map[string]interface{}{
+		"name":   name,
+		"detach": true,
+	}
+
+	var startConfig = map[string]interface{}{
+		"publishAllPorts": false,
+		"privileged":      isTrue(instance, "privileged"),
+		"ReadonlyRootfs":  isTrue(instance, "readOnly"),
+	}
+
+	// These _setupSimpleConfigFields calls should happen before all
+	// other config because they stomp over config fields that other
+	// setup methods might append to. Example: the environment field
+	// setupSimpleConfigFields(createConfig, instance, CreateConfigFields)
+
+	// setupSimpleConfigFields(startConfig, instance, StartConfigFields)
+
+	addLabel(createConfig, map[string]string{UUIDLabel: instance.UUID})
+
+	if len(instanceName) > 0 {
+		addLabel(createConfig, map[string]string{"io.rancher.container.name": instanceName})
+	}
+
+	setupPublishPorts(startConfig, instance)
+
+	setupDNSSearch(startConfig, instance)
+
+	setupLogging(startConfig, instance)
+
+	setupHostname(createConfig, instance)
+
+	setupCommand(createConfig, instance)
+
+	setupPorts(createConfig, instance, startConfig)
+
+	setupVolumes(createConfig, instance, startConfig, client)
+
+	setupLinks(startConfig, instance)
+
+	setupNetworking(instance, host, createConfig, startConfig)
+
+	flagSystemContainer(instance, createConfig)
+
+	setupProxy(instance, createConfig)
+
+	setupCattleConfigURL(instance, createConfig)
+	hostConfig := createHostConfig(startConfig)
```

Comment:
You create startConfig as a map just to decode it to a struct here. What if you just used the HostConfig struct from the beginning? (Same with createConfig)